From 71035700aab22426a95c1a837f7507ab326f81eb Mon Sep 17 00:00:00 2001
From: Chen Jiahao <chenjiahao16@huawei.com>
Date: Sun, 13 Nov 2022 01:26:30 +0800
Subject: [PATCH 0037/1398] [Huawei] ARM64: Introduce
 CONFIG_ARM64_NOSPECTRE_BHB

Offering: HULK
hulk inclusion
category: feature
bugzilla: 186682

-------------------------------

Introduce CONFIG_ARM64_NOSPECTRE_BHB to isolate feature
which providing cmdline option "nospectre_bhb" to disable
Spectre-BHB mitigation on ARM64 for better performance.

Signed-off-by: Chen Jiahao <chenjiahao16@huawei.com>
---
 arch/arm64/Kconfig              | 7 +++++++
 arch/arm64/kernel/proton-pack.c | 9 ++++++++-
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index ff15ea9a27a4..6c249ee0fe12 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -1345,6 +1345,13 @@ config MITIGATE_SPECTRE_BRANCH_HISTORY
 	  When taking an exception from user-space, a sequence of branches
 	  or a firmware call overwrites the branch history.
 
+config ARM64_NOSPECTRE_BHB
+	bool "`nospectre_bhb` cmdline support for 64-bit ARM"
+	default n
+	help
+	  This option allows ARM64 systems to turn off Spectre-bhb mitigation
+	  via `nospectre_bhb` kernel cmdline for better performance.
+
 config RODATA_FULL_DEFAULT_ENABLED
 	bool "Apply r/o permissions of VM areas also to their linear aliases"
 	default y
diff --git a/arch/arm64/kernel/proton-pack.c b/arch/arm64/kernel/proton-pack.c
index a010a28d9b68..b4826fa90f63 100644
--- a/arch/arm64/kernel/proton-pack.c
+++ b/arch/arm64/kernel/proton-pack.c
@@ -81,7 +81,6 @@ ssize_t cpu_show_spectre_v1(struct device *dev, struct device_attribute *attr,
 static enum mitigation_state spectre_v2_state;
 
 static bool __read_mostly __nospectre_v2;
-static bool __read_mostly __nospectre_bhb;
 
 static int __init parse_spectre_v2_param(char *str)
 {
@@ -90,12 +89,16 @@ static int __init parse_spectre_v2_param(char *str)
 }
 early_param("nospectre_v2", parse_spectre_v2_param);
 
+#ifdef CONFIG_ARM64_NOSPECTRE_BHB
+static bool __read_mostly __nospectre_bhb;
+
 static int __init parse_spectre_bhb_param(char *str)
 {
 	__nospectre_bhb = true;
 	return 0;
 }
 early_param("nospectre_bhb", parse_spectre_bhb_param);
+#endif
 
 static bool spectre_v2_mitigations_off(void)
 {
@@ -1069,7 +1072,11 @@ void spectre_bhb_enable_mitigation(const struct arm64_cpu_capabilities *entry)
 		/* No point mitigating Spectre-BHB alone. */
 	} else if (!IS_ENABLED(CONFIG_MITIGATE_SPECTRE_BRANCH_HISTORY)) {
 		pr_info_once("spectre-bhb mitigation disabled by compile time option\n");
+#ifdef CONFIG_ARM64_NOSPECTRE_BHB
 	} else if (__nospectre_bhb || cpu_mitigations_off()) {
+#else
+	} else if (cpu_mitigations_off()) {
+#endif
 		pr_info_once("spectre-bhb mitigation disabled by command line option\n");
 	} else if (supports_ecbhb(SCOPE_LOCAL_CPU)) {
 		state = SPECTRE_MITIGATED;
-- 
2.17.1

